casr/src/bin/casr-cluster.rs

```diff
@@ -343,6 +302,241 @@ fn merge_dirs(input: &Path, output: &Path) -> Result<u64> {
+        }
+
+        // Save casrep
+        added += 1;
+        fs::copy(
+            casrep,
+            format!(
+                "{}/{}",
+                &cluster_dirs[number - 1].display(),
+                &casrep.file_name().unwrap().to_str().unwrap()
+            ),
+        )?;
+
+        // Update cluster
+        let i = clusters.iter().position(|a| a.number == number).unwrap();
+        clusters[i].push(stacktrace.to_vec());
+    }
+
+    // Handle deviant casreps
+    let (result, before, after) = if !deviants.is_empty() {
+        // Copy casrep to tmp dir
+        let deviant_dir = format!("{}/deviant", &oldpath.display());
+        fs::create_dir_all(&deviant_dir)?;
+        for casrep in deviants {
+            fs::copy(
+                casrep,
+                format!(
+                    "{}/{}",
+                    &deviant_dir,
+                    &casrep.file_name().unwrap().to_str().unwrap()
+                ),
+            )?;
+        }
+        // Cluster deviant casreps
+        let (result, before, after) =
+            make_clusters(Path::new(&deviant_dir), Some(oldpath), jobs, dedup, len)?;
+        let _ = fs::remove_dir_all(&deviant_dir);
+        (result, before, after)
+    } else {
+        (0, 0, 0)
+    };
+    Ok((added, duplicates, deduplicated, result, before, after))
+}
+
+/// Calculate silhouette coefficient
+///
+/// # Arguments
+///
+/// * `dir` - path to directory with CASR report clusters
+///
+/// * `jobs` - number of jobs for calculating process
+///
+/// # Return value
+///
+/// Silhouette coefficient
+fn avg_sil(dir: &Path, jobs: usize) -> Result<f64> {
+    // Get cluster dirs
+    let mut dirs: Vec<PathBuf> = fs::read_dir(dir)
+        .unwrap()
+        .map(|path| path.unwrap().path())
+        .filter(|path| {
+            path.clone()
+                .file_name()
+                .unwrap()
+                .to_str()
+                .unwrap()
+                .starts_with("cl")
+        })
+        .collect();
+    dirs.sort();
+
+    if dirs.len() < 2 {
+        bail!("{} valid cluster, nothing to calculate...", dirs.len());
+    }
+
+    // Init clusters vector
+    let mut clusters: Vec<Vec<Stacktrace>> = Vec::new();
+    // Init casreps nuber counter
+    let mut size = 0usize;
+    // Get casreps from each cluster
+    for dir in &dirs {
+        // Get casreps from cluster
+        let casreps = util::get_reports(dir)?;
+        // Get stacktraces from cluster
+        let (_, stacktraces, _, _) = util::reports_from_dirs(casreps, jobs);
+        // Update size
+        size += stacktraces.len();
+        // Add stacktraces
+        clusters.push(stacktraces);
+    }
+    // Init sil sum
+    let mut sum = 0f64;
+    // Calculate silhouette coefficient for each casrep
+    for i in 0..clusters.len() {
+        for num in 0..clusters[i].len() {
+            let sil = sil_coef(num, i, &clusters);
+            sum += sil;
+        }
+    }
+    Ok(sum / size as f64)
```

Comment:
Possible `DivByZero`, if we have not found any valid casrep in clusters. Unlikely, but possible

Reply:
Fixed